// @tsparticles/plugin-interactivity/cjs/InteractionManager.js
import { safeIntersectionObserver, } from "@tsparticles/engine";
import { clickEvent, touchCancelEvent, touchEndEvent, touchMoveEvent, touchStartEvent, } from "./InteractivityConstants.js";
import { InteractivityEventListeners } from "./InteractivityEventListeners.js";
import { InteractorType } from "./Enums/InteractorType.js";
const clickRadius = 1, touchEndLengthOffset = 1, minCoordinate = 0;
export class InteractionManager {
    container;
    interactivityData;
    _clickHandlers;
    _eventListeners;
    _externalInteractors;
    _interactors;
    _intersectionObserver;
    _particleInteractors;
    _pluginManager;
    constructor(pluginManager, container) {
        this.container = container;
        this._pluginManager = pluginManager;
        this._interactors = [];
        this._externalInteractors = [];
        this._particleInteractors = [];
        this._clickHandlers = new Map();
        this._eventListeners = new InteractivityEventListeners(container, this);
        this.interactivityData = {
            mouse: {
                clicking: false,
                inside: false,
            },
        };
        this._intersectionObserver = safeIntersectionObserver(entries => {
            this._intersectionManager(entries);
        });
    }
    addClickHandler(callback) {
        const { container, interactivityData } = this;
        if (container.destroyed) {
            return;
        }
        const el = interactivityData.element;
        if (!el) {
            return;
        }
        const clickOrTouchHandler = (e, pos, radius) => {
            if (container.destroyed) {
                return;
            }
            const pxRatio = container.retina.pixelRatio, posRetina = {
                x: pos.x * pxRatio,
                y: pos.y * pxRatio,
            }, particles = container.particles.grid.queryCircle(posRetina, radius * pxRatio);
            callback(e, particles);
        }, clickHandler = (e) => {
            if (container.destroyed) {
                return;
            }
            const mouseEvent = e, pos = {
                x: mouseEvent.offsetX,
                y: mouseEvent.offsetY,
            };
            clickOrTouchHandler(e, pos, clickRadius);
        }, touchStartHandler = () => {
            if (container.destroyed) {
                return;
            }
            touched = true;
            touchMoved = false;
        }, touchMoveHandler = () => {
            if (container.destroyed) {
                return;
            }
            touchMoved = true;
        }, touchEndHandler = (e) => {
            if (container.destroyed) {
                return;
            }
            if (touched && !touchMoved) {
                const touchEvent = e, lastTouch = touchEvent.touches[touchEvent.touches.length - touchEndLengthOffset];
                if (!lastTouch) {
                    return;
                }
                const element = container.canvas.domElement, canvasRect = element ? element.getBoundingClientRect() : undefined, pos = {
                    x: lastTouch.clientX - (canvasRect ? canvasRect.left : minCoordinate),
                    y: lastTouch.clientY - (canvasRect ? canvasRect.top : minCoordinate),
                };
                clickOrTouchHandler(e, pos, Math.max(lastTouch.radiusX, lastTouch.radiusY));
            }
            touched = false;
            touchMoved = false;
        }, touchCancelHandler = () => {
            if (container.destroyed) {
                return;
            }
            touched = false;
            touchMoved = false;
        };
        let touched = false, touchMoved = false;
        this._clickHandlers.set(clickEvent, clickHandler);
        this._clickHandlers.set(touchStartEvent, touchStartHandler);
        this._clickHandlers.set(touchMoveEvent, touchMoveHandler);
        this._clickHandlers.set(touchEndEvent, touchEndHandler);
        this._clickHandlers.set(touchCancelEvent, touchCancelHandler);
        for (const [key, handler] of this._clickHandlers) {
            el.addEventListener(key, handler);
        }
    }
    addListeners() {
        this._eventListeners.addListeners();
    }
    clearClickHandlers() {
        const { container, interactivityData } = this;
        if (container.destroyed) {
            return;
        }
        for (const [key, handler] of this._clickHandlers) {
            interactivityData.element?.removeEventListener(key, handler);
        }
        this._clickHandlers.clear();
    }
    externalInteract(delta) {
        for (const interactor of this._externalInteractors) {
            const { interactivityData } = this;
            if (interactor.isEnabled(interactivityData)) {
                interactor.interact(interactivityData, delta);
            }
        }
    }
    handleClickMode(mode) {
        if (this.container.destroyed) {
            return;
        }
        const { interactivityData } = this;
        for (const interactor of this._externalInteractors) {
            interactor.handleClickMode?.(mode, interactivityData);
        }
    }
    init() {
        this._eventListeners.init();
        for (const interactor of this._interactors) {
            switch (interactor.type) {
                case InteractorType.external:
                    this._externalInteractors.push(interactor);
                    break;
                case InteractorType.particles:
                    this._particleInteractors.push(interactor);
                    break;
            }
            interactor.init();
        }
    }
    async initInteractors() {
        const interactors = await this._pluginManager.getInteractors?.(this.container, true);
        if (!interactors) {
            return;
        }
        this._interactors = interactors;
        this._externalInteractors = [];
        this._particleInteractors = [];
    }
    particlesInteract(particle, delta) {
        const { interactivityData } = this;
        for (const interactor of this._externalInteractors) {
            interactor.clear(particle, delta);
        }
        for (const interactor of this._particleInteractors) {
            if (interactor.isEnabled(particle, interactivityData)) {
                interactor.interact(particle, interactivityData, delta);
            }
        }
    }
    removeListeners() {
        this._eventListeners.removeListeners();
    }
    reset(particle) {
        const { interactivityData } = this;
        for (const interactor of this._externalInteractors) {
            if (interactor.isEnabled(interactivityData)) {
                interactor.reset(interactivityData, particle);
            }
        }
        for (const interactor of this._particleInteractors) {
            if (interactor.isEnabled(particle, interactivityData)) {
                interactor.reset(interactivityData, particle);
            }
        }
    }
    startObserving() {
        const { interactivityData } = this;
        if (interactivityData.element instanceof HTMLElement && this._intersectionObserver) {
            this._intersectionObserver.observe(interactivityData.element);
        }
    }
    stopObserving() {
        const { interactivityData } = this;
        if (interactivityData.element instanceof HTMLElement && this._intersectionObserver) {
            this._intersectionObserver.unobserve(interactivityData.element);
        }
    }
    updateMaxDistance() {
        let maxTotalDistance = 0;
        for (const interactor of this._interactors) {
            if (interactor.maxDistance > maxTotalDistance) {
                maxTotalDistance = interactor.maxDistance;
            }
        }
        const container = this.container;
        container.particles.grid.setCellSize(maxTotalDistance * container.retina.pixelRatio);
    }
    _intersectionManager = entries => {
        const { container } = this;
        if (container.destroyed || !container.actualOptions.pauseOnOutsideViewport) {
            return;
        }
        for (const entry of entries) {
            if (entry.target !== this.interactivityData.element) {
                continue;
            }
            if (entry.isIntersecting) {
                container.play();
            }
            else {
                container.pause();
            }
        }
    };
}
