// @tsparticles/updater-size/tsparticles.updater.size.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Updater v4.0.4 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.updaters = global.__tsParticlesInternals.updaters || {}, global.__tsParticlesInternals.updaters.size = global.__tsParticlesInternals.updaters.size || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    const minLoops = 0;
    class SizeUpdater {
        _container;
        constructor(container) {
            this._container = container;
        }
        init(particle) {
            const container = this._container, sizeOptions = particle.options.size, sizeAnimation = sizeOptions.animation;
            if (sizeAnimation.enable) {
                particle.size.velocity =
                    (particle.retina.sizeAnimationSpeed / engine.percentDenominator) * container.retina.reduceFactor;
                if (!sizeAnimation.sync) {
                    particle.size.velocity *= engine.getRandom();
                }
            }
        }
        isEnabled(particle) {
            return (!particle.destroyed &&
                !particle.spawning &&
                particle.size.enable &&
                ((particle.size.maxLoops ?? minLoops) <= minLoops ||
                    ((particle.size.maxLoops ?? minLoops) > minLoops &&
                        (particle.size.loops ?? minLoops) < (particle.size.maxLoops ?? minLoops))));
        }
        reset(particle) {
            particle.size.time = 0;
            particle.size.loops = 0;
        }
        update(particle, delta) {
            if (!this.isEnabled(particle)) {
                return;
            }
            engine.updateAnimation(particle, particle.size, true, particle.options.size.animation.destroy, delta);
        }
    }

    async function loadSizeUpdater(engine) {
        engine.checkVersion("4.0.4");
        await engine.pluginManager.register(e => {
            e.pluginManager.addParticleUpdater("size", container => {
                return Promise.resolve(new SizeUpdater(container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadSizeUpdater = loadSizeUpdater;

    exports.loadSizeUpdater = loadSizeUpdater;

}));
Object.assign(globalThis.window || globalThis, { loadSizeUpdater: (globalThis.__tsParticlesInternals.updaters.size || {}).loadSizeUpdater });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
