// @tsparticles/plugin-absorbers/browser/AbsorbersPluginInstance.js
import { executeOnSingleOrMultiple, isArray, } from "@tsparticles/engine";
export class AbsorbersPluginInstance {
    _container;
    _instancesManager;
    constructor(container, instancesManager) {
        this._container = container;
        this._instancesManager = instancesManager;
        this._instancesManager.initContainer(container);
    }
    draw(context) {
        for (const absorber of this._instancesManager.getArray(this._container)) {
            absorber.draw(context);
        }
    }
    async init() {
        const absorbers = this._container.actualOptions.absorbers, promises = executeOnSingleOrMultiple(absorbers, async (absorber) => {
            await this._instancesManager.addAbsorber(this._container, absorber);
        });
        if (isArray(promises)) {
            await Promise.all(promises);
        }
        else {
            await promises;
        }
    }
    particleUpdate(particle, delta) {
        for (const absorber of this._instancesManager.getArray(this._container)) {
            absorber.attract(particle, delta);
            if (particle.destroyed) {
                break;
            }
        }
    }
    resize() {
        for (const absorber of this._instancesManager.getArray(this._container)) {
            absorber.resize();
        }
    }
    stop() {
        this._instancesManager.clear(this._container);
    }
}
