// @tsparticles/engine/cjs/Core/Container.js
import { animate, cancelAnimation, getRangeValue } from "../Utils/MathUtils.js";
import { defaultFps, defaultFpsLimit, millisecondsToSeconds, minFpsLimit } from "./Utils/Constants.js";
import { CanvasManager } from "./CanvasManager.js";
import { EventListeners } from "./Utils/EventListeners.js";
import { EventType } from "../Enums/Types/EventType.js";
import { Options } from "../Options/Classes/Options.js";
import { ParticlesManager } from "./ParticlesManager.js";
import { Retina } from "./Retina.js";
import { getLogger } from "../Utils/LogUtils.js";
import { loadOptions } from "../Utils/OptionsUtils.js";
function guardCheck(container) {
    return !container.destroyed;
}
function updateDelta(delta, value, fpsLimit = defaultFps, smooth = false) {
    delta.value = value;
    delta.factor = smooth ? defaultFps / fpsLimit : (defaultFps * value) / millisecondsToSeconds;
}
function loadContainerOptions(pluginManager, container, ...sourceOptionsArr) {
    const options = new Options(pluginManager, container);
    loadOptions(options, ...sourceOptionsArr);
    return options;
}
export class Container {
    actualOptions;
    canvas;
    destroyed;
    effectDrawers;
    fpsLimit;
    hdr;
    id;
    pageHidden;
    particleCreatedPlugins;
    particleDestroyedPlugins;
    particlePositionPlugins;
    particleUpdaters;
    particles;
    plugins;
    retina;
    shapeDrawers;
    started;
    zLayers;
    _delay;
    _delayTimeout;
    _delta = { value: 0, factor: 0 };
    _dispatchCallback;
    _drawAnimationFrame;
    _duration;
    _eventListeners;
    _firstStart;
    _initialSourceOptions;
    _lastFrameTime;
    _lifeTime;
    _onDestroy;
    _options;
    _paused;
    _pluginManager;
    _smooth;
    _sourceOptions;
    constructor(params) {
        const { dispatchCallback, pluginManager, id, onDestroy, sourceOptions } = params;
        this._pluginManager = pluginManager;
        this._dispatchCallback = dispatchCallback;
        this._onDestroy = onDestroy;
        this.id = Symbol(id);
        this.fpsLimit = 120;
        this.hdr = false;
        this._smooth = false;
        this._delay = 0;
        this._duration = 0;
        this._lifeTime = 0;
        this._firstStart = true;
        this.started = false;
        this.destroyed = false;
        this._paused = true;
        this._lastFrameTime = 0;
        this.zLayers = 100;
        this.pageHidden = false;
        this._sourceOptions = sourceOptions;
        this._initialSourceOptions = sourceOptions;
        this.effectDrawers = new Map();
        this.shapeDrawers = new Map();
        this.particleUpdaters = [];
        this.retina = new Retina(this);
        this.canvas = new CanvasManager(this._pluginManager, this);
        this.particles = new ParticlesManager(this._pluginManager, this);
        this.plugins = [];
        this.particleDestroyedPlugins = [];
        this.particleCreatedPlugins = [];
        this.particlePositionPlugins = [];
        this._options = loadContainerOptions(this._pluginManager, this);
        this.actualOptions = loadContainerOptions(this._pluginManager, this);
        this._eventListeners = new EventListeners(this);
        this.dispatchEvent(EventType.containerBuilt);
    }
    get animationStatus() {
        return !this._paused && !this.pageHidden && guardCheck(this);
    }
    get options() {
        return this._options;
    }
    get sourceOptions() {
        return this._sourceOptions;
    }
    addLifeTime(value) {
        this._lifeTime += value;
    }
    alive() {
        return !this._duration || this._lifeTime <= this._duration;
    }
    destroy(remove = true) {
        if (!guardCheck(this)) {
            return;
        }
        this.stop();
        this.particles.destroy();
        this.canvas.destroy();
        for (const [, effectDrawer] of this.effectDrawers) {
            effectDrawer.destroy?.(this);
        }
        for (const [, shapeDrawer] of this.shapeDrawers) {
            shapeDrawer.destroy?.(this);
        }
        for (const plugin of this.plugins) {
            plugin.destroy?.();
        }
        this.effectDrawers = new Map();
        this.shapeDrawers = new Map();
        this.particleUpdaters = [];
        this.plugins.length = 0;
        this._pluginManager.clearPlugins(this);
        this.destroyed = true;
        this._onDestroy(remove);
        this.dispatchEvent(EventType.containerDestroyed);
    }
    dispatchEvent(type, data) {
        this._dispatchCallback(type, {
            container: this,
            data,
        });
    }
    draw(force) {
        if (!guardCheck(this)) {
            return;
        }
        let refreshTime = force;
        this._drawAnimationFrame = animate((timestamp) => {
            if (refreshTime) {
                this._lastFrameTime = undefined;
                refreshTime = false;
            }
            this._nextFrame(timestamp);
        });
    }
    async export(type, options = {}) {
        for (const plugin of this.plugins) {
            if (!plugin.export) {
                continue;
            }
            const res = await plugin.export(type, options);
            if (!res.supported) {
                continue;
            }
            return res.blob;
        }
        getLogger().error(`Export plugin with type ${type} not found`);
        return undefined;
    }
    async init() {
        if (!guardCheck(this)) {
            return;
        }
        const allContainerPlugins = new Map();
        for (const plugin of this._pluginManager.plugins) {
            const containerPlugin = await plugin.getPlugin(this);
            if (containerPlugin.preInit) {
                await containerPlugin.preInit();
            }
            allContainerPlugins.set(plugin, containerPlugin);
        }
        await this.initDrawersAndUpdaters();
        this._options = loadContainerOptions(this._pluginManager, this, this._initialSourceOptions, this.sourceOptions);
        this.actualOptions = loadContainerOptions(this._pluginManager, this, this._options);
        this.plugins.length = 0;
        this.particleDestroyedPlugins.length = 0;
        this.particleCreatedPlugins.length = 0;
        this.particlePositionPlugins.length = 0;
        for (const [plugin, containerPlugin] of allContainerPlugins) {
            if (plugin.needsPlugin(this.actualOptions)) {
                this.plugins.push(containerPlugin);
                if (containerPlugin.particleCreated) {
                    this.particleCreatedPlugins.push(containerPlugin);
                }
                if (containerPlugin.particleDestroyed) {
                    this.particleDestroyedPlugins.push(containerPlugin);
                }
                if (containerPlugin.particlePosition) {
                    this.particlePositionPlugins.push(containerPlugin);
                }
            }
        }
        this.retina.init();
        this.canvas.init();
        this.updateActualOptions();
        this.canvas.initBackground();
        this.canvas.resize();
        const { delay, duration, fpsLimit, hdr, smooth, zLayers } = this.actualOptions;
        this.hdr = hdr;
        this.zLayers = zLayers;
        this._duration = getRangeValue(duration) * millisecondsToSeconds;
        this._delay = getRangeValue(delay) * millisecondsToSeconds;
        this._lifeTime = 0;
        this.fpsLimit = fpsLimit > minFpsLimit ? fpsLimit : defaultFpsLimit;
        this._smooth = smooth;
        for (const plugin of this.plugins) {
            await plugin.init?.();
        }
        await this.particles.init();
        this.dispatchEvent(EventType.containerInit);
        this.particles.setDensity();
        for (const plugin of this.plugins) {
            plugin.particlesSetup?.();
        }
        this.dispatchEvent(EventType.particlesSetup);
    }
    async initDrawersAndUpdaters() {
        const pluginManager = this._pluginManager;
        this.effectDrawers = await pluginManager.getEffectDrawers(this, true);
        this.shapeDrawers = await pluginManager.getShapeDrawers(this, true);
        this.particleUpdaters = await pluginManager.getUpdaters(this, true);
    }
    pause() {
        if (!guardCheck(this)) {
            return;
        }
        if (this._drawAnimationFrame !== undefined) {
            cancelAnimation(this._drawAnimationFrame);
            delete this._drawAnimationFrame;
        }
        if (this._paused) {
            return;
        }
        for (const plugin of this.plugins) {
            plugin.pause?.();
        }
        if (!this.pageHidden) {
            this._paused = true;
        }
        this.dispatchEvent(EventType.containerPaused);
    }
    play(force) {
        if (!guardCheck(this)) {
            return;
        }
        const needsUpdate = this._paused || force;
        if (this._firstStart && !this.actualOptions.autoPlay) {
            this._firstStart = false;
            return;
        }
        if (this._paused) {
            this._paused = false;
        }
        if (needsUpdate) {
            for (const plugin of this.plugins) {
                if (plugin.play) {
                    plugin.play();
                }
            }
        }
        this.dispatchEvent(EventType.containerPlay);
        this.draw(needsUpdate ?? false);
    }
    async refresh() {
        if (!guardCheck(this)) {
            return;
        }
        this.stop();
        return this.start();
    }
    async reset(sourceOptions) {
        if (!guardCheck(this)) {
            return;
        }
        this._initialSourceOptions = sourceOptions;
        this._sourceOptions = sourceOptions;
        this._options = loadContainerOptions(this._pluginManager, this, this._initialSourceOptions, this.sourceOptions);
        this.actualOptions = loadContainerOptions(this._pluginManager, this, this._options);
        return this.refresh();
    }
    async start() {
        if (!guardCheck(this) || this.started) {
            return;
        }
        await this.init();
        this.started = true;
        await new Promise(resolve => {
            const start = async () => {
                this._eventListeners.addListeners();
                for (const plugin of this.plugins) {
                    await plugin.start?.();
                }
                this.dispatchEvent(EventType.containerStarted);
                this.play();
                resolve();
            };
            this._delayTimeout = setTimeout(() => void start(), this._delay);
        });
    }
    stop() {
        if (!guardCheck(this) || !this.started) {
            return;
        }
        if (this._delayTimeout) {
            clearTimeout(this._delayTimeout);
            delete this._delayTimeout;
        }
        this._firstStart = true;
        this.started = false;
        this._eventListeners.removeListeners();
        this.pause();
        this.particles.clear();
        this.canvas.stop();
        for (const plugin of this.plugins) {
            plugin.stop?.();
        }
        this.particleCreatedPlugins.length = 0;
        this.particleDestroyedPlugins.length = 0;
        this.particlePositionPlugins.length = 0;
        this._sourceOptions = this._options;
        this.dispatchEvent(EventType.containerStopped);
    }
    updateActualOptions() {
        let refresh = false;
        for (const plugin of this.plugins) {
            if (plugin.updateActualOptions) {
                refresh = plugin.updateActualOptions() || refresh;
            }
        }
        return refresh;
    }
    _nextFrame = (timestamp) => {
        try {
            if (!this._smooth &&
                this._lastFrameTime !== undefined &&
                timestamp < this._lastFrameTime + millisecondsToSeconds / this.fpsLimit) {
                this.draw(false);
                return;
            }
            this._lastFrameTime ??= timestamp;
            updateDelta(this._delta, timestamp - this._lastFrameTime, this.fpsLimit, this._smooth);
            this.addLifeTime(this._delta.value);
            this._lastFrameTime = timestamp;
            if (this._delta.value > millisecondsToSeconds) {
                this.draw(false);
                return;
            }
            this.canvas.render.drawParticles(this._delta);
            if (!this.alive()) {
                this.destroy();
                return;
            }
            if (this.animationStatus) {
                this.draw(false);
            }
        }
        catch (e) {
            getLogger().error("error in animation loop", e);
        }
    };
}
