// @tsparticles/engine/cjs/Core/RenderManager.js
import { clear, drawParticle, drawParticlePlugin, paintBase, paintImage } from "../Utils/CanvasUtils.js";
import { defaultCompositeValue, defaultTransformValue, minimumSize, zIndexFactorOffset } from "./Utils/Constants.js";
import { getStyleFromHsl, rangeColorToHsl } from "../Utils/ColorUtils.js";
const fColorIndex = 0, sColorIndex = 1;
function setTransformValue(factor, newFactor, key) {
    const newValue = newFactor[key];
    if (newValue !== undefined) {
        factor[key] = (factor[key] ?? defaultTransformValue) * newValue;
    }
}
export class RenderManager {
    _canvasClearPlugins;
    _canvasManager;
    _canvasPaintPlugins;
    _clearDrawPlugins;
    _colorPlugins;
    _container;
    _context;
    _contextSettings;
    _drawParticlePlugins;
    _drawParticlesCleanupPlugins;
    _drawParticlesSetupPlugins;
    _drawPlugins;
    _drawSettingsCleanupPlugins;
    _drawSettingsSetupPlugins;
    _pluginManager;
    _postDrawUpdaters;
    _preDrawUpdaters;
    _reusableColorStyles = {};
    _reusablePluginColors = [undefined, undefined];
    _reusableTransform = {};
    constructor(pluginManager, container, canvasManager) {
        this._pluginManager = pluginManager;
        this._container = container;
        this._canvasManager = canvasManager;
        this._context = null;
        this._preDrawUpdaters = [];
        this._postDrawUpdaters = [];
        this._colorPlugins = [];
        this._canvasClearPlugins = [];
        this._canvasPaintPlugins = [];
        this._clearDrawPlugins = [];
        this._drawParticlePlugins = [];
        this._drawParticlesCleanupPlugins = [];
        this._drawParticlesSetupPlugins = [];
        this._drawPlugins = [];
        this._drawSettingsSetupPlugins = [];
        this._drawSettingsCleanupPlugins = [];
    }
    get settings() {
        return this._contextSettings;
    }
    canvasClear() {
        if (!this._container.actualOptions.clear) {
            return;
        }
        this.draw(ctx => {
            clear(ctx, this._canvasManager.size);
        });
    }
    clear() {
        let pluginHandled = false;
        for (const plugin of this._canvasClearPlugins) {
            pluginHandled = plugin.canvasClear?.() ?? false;
            if (pluginHandled) {
                break;
            }
        }
        if (pluginHandled) {
            return;
        }
        this.canvasClear();
    }
    destroy() {
        this.stop();
        this._preDrawUpdaters = [];
        this._postDrawUpdaters = [];
        this._colorPlugins = [];
        this._canvasClearPlugins = [];
        this._canvasPaintPlugins = [];
        this._clearDrawPlugins = [];
        this._drawParticlePlugins = [];
        this._drawParticlesCleanupPlugins = [];
        this._drawParticlesSetupPlugins = [];
        this._drawPlugins = [];
        this._drawSettingsSetupPlugins = [];
        this._drawSettingsCleanupPlugins = [];
    }
    draw(cb) {
        const ctx = this._context;
        if (!ctx) {
            return;
        }
        return cb(ctx);
    }
    drawParticle(particle, delta) {
        if (particle.spawning || particle.destroyed) {
            return;
        }
        const radius = particle.getRadius();
        if (radius <= minimumSize) {
            return;
        }
        const pfColor = particle.getFillColor(), psColor = particle.getStrokeColor();
        let [fColor, sColor] = this._getPluginParticleColors(particle);
        fColor ??= pfColor;
        sColor ??= psColor;
        if (!fColor && !sColor) {
            return;
        }
        const container = this._container, zIndexOptions = particle.options.zIndex, zIndexFactor = zIndexFactorOffset - particle.zIndexFactor, { fillOpacity, opacity, strokeOpacity } = particle.getOpacity(), transform = this._reusableTransform, colorStyles = this._reusableColorStyles, fill = fColor ? getStyleFromHsl(fColor, container.hdr, fillOpacity * opacity) : undefined, stroke = sColor ? getStyleFromHsl(sColor, container.hdr, strokeOpacity * opacity) : fill;
        transform.a = transform.b = transform.c = transform.d = undefined;
        colorStyles.fill = fill;
        colorStyles.stroke = stroke;
        this.draw((context) => {
            for (const plugin of this._drawParticlesSetupPlugins) {
                plugin.drawParticleSetup?.(context, particle, delta);
            }
            this._applyPreDrawUpdaters(context, particle, radius, opacity, colorStyles, transform);
            drawParticle({
                container,
                context,
                particle,
                delta,
                colorStyles,
                radius: radius * zIndexFactor ** zIndexOptions.sizeRate,
                opacity: opacity,
                transform,
            });
            this._applyPostDrawUpdaters(particle);
            for (const plugin of this._drawParticlesCleanupPlugins) {
                plugin.drawParticleCleanup?.(context, particle, delta);
            }
        });
    }
    drawParticlePlugins(particle, delta) {
        this.draw(ctx => {
            for (const plugin of this._drawParticlePlugins) {
                drawParticlePlugin(ctx, plugin, particle, delta);
            }
        });
    }
    drawParticles(delta) {
        const { particles } = this._container;
        this.clear();
        particles.update(delta);
        this.draw(ctx => {
            for (const plugin of this._drawSettingsSetupPlugins) {
                plugin.drawSettingsSetup?.(ctx, delta);
            }
            for (const plugin of this._drawPlugins) {
                plugin.draw?.(ctx, delta);
            }
            particles.drawParticles(delta);
            for (const plugin of this._clearDrawPlugins) {
                plugin.clearDraw?.(ctx, delta);
            }
            for (const plugin of this._drawSettingsCleanupPlugins) {
                plugin.drawSettingsCleanup?.(ctx, delta);
            }
        });
    }
    init() {
        this.initUpdaters();
        this.initPlugins();
        this.paint();
    }
    initPlugins() {
        this._colorPlugins = [];
        this._canvasClearPlugins = [];
        this._canvasPaintPlugins = [];
        this._clearDrawPlugins = [];
        this._drawParticlePlugins = [];
        this._drawParticlesSetupPlugins = [];
        this._drawParticlesCleanupPlugins = [];
        this._drawPlugins = [];
        this._drawSettingsSetupPlugins = [];
        this._drawSettingsCleanupPlugins = [];
        for (const plugin of this._container.plugins) {
            if (plugin.particleFillColor ?? plugin.particleStrokeColor) {
                this._colorPlugins.push(plugin);
            }
            if (plugin.canvasClear) {
                this._canvasClearPlugins.push(plugin);
            }
            if (plugin.canvasPaint) {
                this._canvasPaintPlugins.push(plugin);
            }
            if (plugin.drawParticle) {
                this._drawParticlePlugins.push(plugin);
            }
            if (plugin.drawParticleSetup) {
                this._drawParticlesSetupPlugins.push(plugin);
            }
            if (plugin.drawParticleCleanup) {
                this._drawParticlesCleanupPlugins.push(plugin);
            }
            if (plugin.draw) {
                this._drawPlugins.push(plugin);
            }
            if (plugin.drawSettingsSetup) {
                this._drawSettingsSetupPlugins.push(plugin);
            }
            if (plugin.drawSettingsCleanup) {
                this._drawSettingsCleanupPlugins.push(plugin);
            }
            if (plugin.clearDraw) {
                this._clearDrawPlugins.push(plugin);
            }
        }
    }
    initUpdaters() {
        this._preDrawUpdaters = [];
        this._postDrawUpdaters = [];
        for (const updater of this._container.particleUpdaters) {
            if (updater.afterDraw) {
                this._postDrawUpdaters.push(updater);
            }
            if (updater.getColorStyles ?? updater.getTransformValues ?? updater.beforeDraw) {
                this._preDrawUpdaters.push(updater);
            }
        }
    }
    paint() {
        let handled = false;
        for (const plugin of this._canvasPaintPlugins) {
            handled = plugin.canvasPaint?.() ?? false;
            if (handled) {
                break;
            }
        }
        if (handled) {
            return;
        }
        this.paintBase();
    }
    paintBase(baseColor) {
        this.draw(ctx => {
            paintBase(ctx, this._canvasManager.size, baseColor);
        });
    }
    paintImage(image, opacity) {
        this.draw(ctx => {
            paintImage(ctx, this._canvasManager.size, image, opacity);
        });
    }
    setContext(context) {
        this._context = context;
        if (this._context) {
            this._context.globalCompositeOperation = defaultCompositeValue;
        }
    }
    setContextSettings(settings) {
        this._contextSettings = settings;
    }
    stop() {
        this.draw(ctx => {
            clear(ctx, this._canvasManager.size);
        });
    }
    _applyPostDrawUpdaters = particle => {
        for (const updater of this._postDrawUpdaters) {
            updater.afterDraw?.(particle);
        }
    };
    _applyPreDrawUpdaters = (ctx, particle, radius, zOpacity, colorStyles, transform) => {
        for (const updater of this._preDrawUpdaters) {
            if (updater.getColorStyles) {
                const { fill, stroke } = updater.getColorStyles(particle, ctx, radius, zOpacity);
                if (fill) {
                    colorStyles.fill = fill;
                }
                if (stroke) {
                    colorStyles.stroke = stroke;
                }
            }
            if (updater.getTransformValues) {
                const updaterTransform = updater.getTransformValues(particle);
                for (const key in updaterTransform) {
                    setTransformValue(transform, updaterTransform, key);
                }
            }
            updater.beforeDraw?.(particle);
        }
    };
    _getPluginParticleColors = particle => {
        let fColor, sColor;
        for (const plugin of this._colorPlugins) {
            if (!fColor && plugin.particleFillColor) {
                fColor = rangeColorToHsl(this._pluginManager, plugin.particleFillColor(particle));
            }
            if (!sColor && plugin.particleStrokeColor) {
                sColor = rangeColorToHsl(this._pluginManager, plugin.particleStrokeColor(particle));
            }
            if (fColor && sColor) {
                break;
            }
        }
        this._reusablePluginColors[fColorIndex] = fColor;
        this._reusablePluginColors[sColorIndex] = sColor;
        return this._reusablePluginColors;
    };
}
