// @tsparticles/interaction-external-drag/browser/Dragger.js
import { ExternalInteractorBase, } from "@tsparticles/plugin-interactivity";
import { defaultFps, getDistance, isInArray, millisecondsToSeconds, } from "@tsparticles/engine";
import { Drag } from "./Options/Classes/Drag.js";
const clickReleaseDelayMs = 50, defaultMomentumFactor = 0.03, dragMode = "drag", firstSampleIndex = 0, lastSampleOffset = 1, maxMomentumSamples = 5, minSamplesForMomentum = 2, momentumToVelocityFactor = millisecondsToSeconds / defaultFps, zeroValue = 0;
export class Dragger extends ExternalInteractorBase {
    maxDistance = 0;
    _dragStartClickTime;
    _dragStartClickingFalseTime;
    _draggedParticle;
    _grabOffset;
    _lastMousePosition;
    _momentumSamples = [];
    _mouseDownHandled = false;
    _savedVelocity;
    constructor(container) {
        super(container);
    }
    clear(_particle, _delta) {
    }
    init() {
    }
    interact(interactivityData, _delta) {
        const mouse = interactivityData.mouse, mousePos = mouse.position;
        if (mousePos) {
            this._lastMousePosition = mousePos;
        }
        if (this._draggedParticle) {
            const now = performance.now();
            if (!mouse.clicking) {
                this._dragStartClickingFalseTime ??= now;
            }
            else {
                this._dragStartClickingFalseTime = undefined;
            }
            const hasStableClickEndWithoutClickTime = this._dragStartClickTime === undefined &&
                this._dragStartClickingFalseTime !== undefined &&
                now - this._dragStartClickingFalseTime >= clickReleaseDelayMs;
            if (!mouse.inside || this._hasClickEnded(mouse) || hasStableClickEndWithoutClickTime) {
                this._releaseDragged();
                this._mouseDownHandled = false;
                return;
            }
            const pointerPos = mousePos ?? this._lastMousePosition;
            if (!pointerPos) {
                return;
            }
            this._momentumSamples.push({ x: pointerPos.x, y: pointerPos.y, t: now });
            if (this._momentumSamples.length > maxMomentumSamples) {
                this._momentumSamples.shift();
            }
            const targetX = pointerPos.x + (this._grabOffset?.x ?? zeroValue), targetY = pointerPos.y + (this._grabOffset?.y ?? zeroValue);
            this._draggedParticle.position.x = targetX - this._draggedParticle.offset.x;
            this._draggedParticle.position.y = targetY - this._draggedParticle.offset.y;
            this._draggedParticle.velocity.x = zeroValue;
            this._draggedParticle.velocity.y = zeroValue;
            this._draggedParticle.initialPosition.x = this._draggedParticle.position.x;
            this._draggedParticle.initialPosition.y = this._draggedParticle.position.y;
            this._draggedParticle.misplaced = false;
            return;
        }
        if (!mouse.clicking) {
            this._releaseDragged();
            this._mouseDownHandled = false;
            return;
        }
        if (this._mouseDownHandled) {
            return;
        }
        this._mouseDownHandled = true;
        const downPos = mouse.downPosition ?? mousePos ?? this._lastMousePosition, closest = this._findParticleUnderCursor(interactivityData, downPos);
        if (!closest || !mousePos) {
            if (!closest) {
                this._mouseDownHandled = false;
            }
            return;
        }
        const renderedPos = closest.getPosition();
        this._draggedParticle = closest;
        this._dragStartClickTime = mouse.clickTime;
        this._grabOffset = {
            x: renderedPos.x - mousePos.x,
            y: renderedPos.y - mousePos.y,
        };
        this._savedVelocity = { x: closest.velocity.x, y: closest.velocity.y };
        this._dragStartClickingFalseTime = undefined;
        this._momentumSamples = [{ x: mousePos.x, y: mousePos.y, t: performance.now() }];
        closest.velocity.x = zeroValue;
        closest.velocity.y = zeroValue;
        closest.position.x = renderedPos.x - closest.offset.x;
        closest.position.y = renderedPos.y - closest.offset.y;
        closest.initialPosition.x = closest.position.x;
        closest.initialPosition.y = closest.position.y;
        closest.misplaced = false;
    }
    isEnabled(interactivityData, particle) {
        const container = this.container, mouse = interactivityData.mouse, events = (particle?.interactivity ?? container.actualOptions.interactivity)?.events;
        if (this._draggedParticle || this._mouseDownHandled) {
            return true;
        }
        return !!events?.onClick.enable && mouse.clicking && !!mouse.position && isInArray(dragMode, events.onClick.mode);
    }
    loadModeOptions(options, ...sources) {
        options.drag ??= new Drag();
        for (const source of sources) {
            options.drag.load(source?.drag);
        }
    }
    reset(_interactivityData, _particle) {
    }
    _findParticleUnderCursor(interactivityData, mousePos) {
        if (!mousePos) {
            return;
        }
        const candidates = this.container.particles.filter(p => this.isEnabled(interactivityData, p));
        let closest, closestDist = Infinity;
        for (const candidate of candidates) {
            const dist = getDistance(candidate.getPosition(), mousePos), radius = candidate.getRadius();
            if (!Number.isFinite(dist) || !Number.isFinite(radius) || dist > radius || dist >= closestDist) {
                continue;
            }
            closest = candidate;
            closestDist = dist;
        }
        return closest;
    }
    _hasClickEnded(mouse) {
        return (this._dragStartClickTime !== undefined &&
            mouse.clickTime !== undefined &&
            mouse.clickTime !== this._dragStartClickTime);
    }
    _releaseDragged() {
        if (this._draggedParticle) {
            const dragOptions = this.container.actualOptions.interactivity?.modes.drag, preserveMomentum = dragOptions?.preserveMomentum ?? false;
            if (preserveMomentum && this._momentumSamples.length >= minSamplesForMomentum) {
                const first = this._momentumSamples[firstSampleIndex], last = this._momentumSamples[this._momentumSamples.length - lastSampleOffset];
                if (first && last) {
                    const dt = last.t - first.t;
                    if (dt > zeroValue) {
                        const factor = dragOptions?.momentumFactor ?? defaultMomentumFactor;
                        this._draggedParticle.velocity.x = ((last.x - first.x) / dt) * momentumToVelocityFactor * factor;
                        this._draggedParticle.velocity.y = ((last.y - first.y) / dt) * momentumToVelocityFactor * factor;
                    }
                    else if (this._savedVelocity) {
                        this._draggedParticle.velocity.x = this._savedVelocity.x;
                        this._draggedParticle.velocity.y = this._savedVelocity.y;
                    }
                }
                else if (this._savedVelocity) {
                    this._draggedParticle.velocity.x = this._savedVelocity.x;
                    this._draggedParticle.velocity.y = this._savedVelocity.y;
                }
            }
            else if (this._savedVelocity) {
                this._draggedParticle.velocity.x = this._savedVelocity.x;
                this._draggedParticle.velocity.y = this._savedVelocity.y;
            }
        }
        this._draggedParticle = undefined;
        this._dragStartClickTime = undefined;
        this._dragStartClickingFalseTime = undefined;
        this._grabOffset = undefined;
        this._lastMousePosition = undefined;
        this._momentumSamples = [];
        this._savedVelocity = undefined;
    }
}
