// @tsparticles/interaction-external-drag/tsparticles.interaction.external.drag.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* External Interaction v4.0.4 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/plugin-interactivity'), require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/plugin-interactivity', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.interactions = global.__tsParticlesInternals.interactions || {}, global.__tsParticlesInternals.interactions.externalDrag = global.__tsParticlesInternals.interactions.externalDrag || {}), global.__tsParticlesInternals.plugins.interactivity, global.__tsParticlesInternals.engine));
})(this, (function (exports, pluginInteractivity, engine) { 'use strict';

    const defaultMomentumFactor$1 = 0.03, defaultPreserveMomentum = false;
    class Drag {
        momentumFactor;
        preserveMomentum;
        constructor() {
            this.momentumFactor = defaultMomentumFactor$1;
            this.preserveMomentum = defaultPreserveMomentum;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.momentumFactor !== undefined) {
                this.momentumFactor = data.momentumFactor;
            }
            if (data.preserveMomentum !== undefined) {
                this.preserveMomentum = data.preserveMomentum;
            }
        }
    }

    const clickReleaseDelayMs = 50, defaultMomentumFactor = 0.03, dragMode = "drag", firstSampleIndex = 0, lastSampleOffset = 1, maxMomentumSamples = 5, minSamplesForMomentum = 2, momentumToVelocityFactor = engine.millisecondsToSeconds / engine.defaultFps, zeroValue = 0;
    class Dragger extends pluginInteractivity.ExternalInteractorBase {
        maxDistance = 0;
        _dragStartClickTime;
        _dragStartClickingFalseTime;
        _draggedParticle;
        _grabOffset;
        _lastMousePosition;
        _momentumSamples = [];
        _mouseDownHandled = false;
        _savedVelocity;
        constructor(container) {
            super(container);
        }
        clear(_particle, _delta) {
        }
        init() {
        }
        interact(interactivityData, _delta) {
            const mouse = interactivityData.mouse, mousePos = mouse.position;
            if (mousePos) {
                this._lastMousePosition = mousePos;
            }
            if (this._draggedParticle) {
                const now = performance.now();
                if (!mouse.clicking) {
                    this._dragStartClickingFalseTime ??= now;
                }
                else {
                    this._dragStartClickingFalseTime = undefined;
                }
                const hasStableClickEndWithoutClickTime = this._dragStartClickTime === undefined &&
                    this._dragStartClickingFalseTime !== undefined &&
                    now - this._dragStartClickingFalseTime >= clickReleaseDelayMs;
                if (!mouse.inside || this._hasClickEnded(mouse) || hasStableClickEndWithoutClickTime) {
                    this._releaseDragged();
                    this._mouseDownHandled = false;
                    return;
                }
                const pointerPos = mousePos ?? this._lastMousePosition;
                if (!pointerPos) {
                    return;
                }
                this._momentumSamples.push({ x: pointerPos.x, y: pointerPos.y, t: now });
                if (this._momentumSamples.length > maxMomentumSamples) {
                    this._momentumSamples.shift();
                }
                const targetX = pointerPos.x + (this._grabOffset?.x ?? zeroValue), targetY = pointerPos.y + (this._grabOffset?.y ?? zeroValue);
                this._draggedParticle.position.x = targetX - this._draggedParticle.offset.x;
                this._draggedParticle.position.y = targetY - this._draggedParticle.offset.y;
                this._draggedParticle.velocity.x = zeroValue;
                this._draggedParticle.velocity.y = zeroValue;
                this._draggedParticle.initialPosition.x = this._draggedParticle.position.x;
                this._draggedParticle.initialPosition.y = this._draggedParticle.position.y;
                this._draggedParticle.misplaced = false;
                return;
            }
            if (!mouse.clicking) {
                this._releaseDragged();
                this._mouseDownHandled = false;
                return;
            }
            if (this._mouseDownHandled) {
                return;
            }
            this._mouseDownHandled = true;
            const downPos = mouse.downPosition ?? mousePos ?? this._lastMousePosition, closest = this._findParticleUnderCursor(interactivityData, downPos);
            if (!closest || !mousePos) {
                if (!closest) {
                    this._mouseDownHandled = false;
                }
                return;
            }
            const renderedPos = closest.getPosition();
            this._draggedParticle = closest;
            this._dragStartClickTime = mouse.clickTime;
            this._grabOffset = {
                x: renderedPos.x - mousePos.x,
                y: renderedPos.y - mousePos.y,
            };
            this._savedVelocity = { x: closest.velocity.x, y: closest.velocity.y };
            this._dragStartClickingFalseTime = undefined;
            this._momentumSamples = [{ x: mousePos.x, y: mousePos.y, t: performance.now() }];
            closest.velocity.x = zeroValue;
            closest.velocity.y = zeroValue;
            closest.position.x = renderedPos.x - closest.offset.x;
            closest.position.y = renderedPos.y - closest.offset.y;
            closest.initialPosition.x = closest.position.x;
            closest.initialPosition.y = closest.position.y;
            closest.misplaced = false;
        }
        isEnabled(interactivityData, particle) {
            const container = this.container, mouse = interactivityData.mouse, events = (particle?.interactivity ?? container.actualOptions.interactivity)?.events;
            if (this._draggedParticle || this._mouseDownHandled) {
                return true;
            }
            return !!events?.onClick.enable && mouse.clicking && !!mouse.position && engine.isInArray(dragMode, events.onClick.mode);
        }
        loadModeOptions(options, ...sources) {
            options.drag ??= new Drag();
            for (const source of sources) {
                options.drag.load(source?.drag);
            }
        }
        reset(_interactivityData, _particle) {
        }
        _findParticleUnderCursor(interactivityData, mousePos) {
            if (!mousePos) {
                return;
            }
            const candidates = this.container.particles.filter(p => this.isEnabled(interactivityData, p));
            let closest, closestDist = Infinity;
            for (const candidate of candidates) {
                const dist = engine.getDistance(candidate.getPosition(), mousePos), radius = candidate.getRadius();
                if (!Number.isFinite(dist) || !Number.isFinite(radius) || dist > radius || dist >= closestDist) {
                    continue;
                }
                closest = candidate;
                closestDist = dist;
            }
            return closest;
        }
        _hasClickEnded(mouse) {
            return (this._dragStartClickTime !== undefined &&
                mouse.clickTime !== undefined &&
                mouse.clickTime !== this._dragStartClickTime);
        }
        _releaseDragged() {
            if (this._draggedParticle) {
                const dragOptions = this.container.actualOptions.interactivity?.modes.drag, preserveMomentum = dragOptions?.preserveMomentum ?? false;
                if (preserveMomentum && this._momentumSamples.length >= minSamplesForMomentum) {
                    const first = this._momentumSamples[firstSampleIndex], last = this._momentumSamples[this._momentumSamples.length - lastSampleOffset];
                    if (first && last) {
                        const dt = last.t - first.t;
                        if (dt > zeroValue) {
                            const factor = dragOptions?.momentumFactor ?? defaultMomentumFactor;
                            this._draggedParticle.velocity.x = ((last.x - first.x) / dt) * momentumToVelocityFactor * factor;
                            this._draggedParticle.velocity.y = ((last.y - first.y) / dt) * momentumToVelocityFactor * factor;
                        }
                        else if (this._savedVelocity) {
                            this._draggedParticle.velocity.x = this._savedVelocity.x;
                            this._draggedParticle.velocity.y = this._savedVelocity.y;
                        }
                    }
                    else if (this._savedVelocity) {
                        this._draggedParticle.velocity.x = this._savedVelocity.x;
                        this._draggedParticle.velocity.y = this._savedVelocity.y;
                    }
                }
                else if (this._savedVelocity) {
                    this._draggedParticle.velocity.x = this._savedVelocity.x;
                    this._draggedParticle.velocity.y = this._savedVelocity.y;
                }
            }
            this._draggedParticle = undefined;
            this._dragStartClickTime = undefined;
            this._dragStartClickingFalseTime = undefined;
            this._grabOffset = undefined;
            this._lastMousePosition = undefined;
            this._momentumSamples = [];
            this._savedVelocity = undefined;
        }
    }

    async function loadExternalDragInteraction(engine) {
        engine.checkVersion("4.0.4");
        await engine.pluginManager.register((e) => {
            pluginInteractivity.ensureInteractivityPluginLoaded(e);
            e.pluginManager.addInteractor?.("externalDrag", container => {
                return Promise.resolve(new Dragger(container));
            });
        });
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadExternalDragInteraction = loadExternalDragInteraction;

    exports.Drag = Drag;
    exports.loadExternalDragInteraction = loadExternalDragInteraction;

}));
Object.assign(globalThis.window || globalThis, { loadExternalDragInteraction: (globalThis.__tsParticlesInternals.interactions.externalDrag || {}).loadExternalDragInteraction });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
