// @tsparticles/engine/esm/Core/ParticlesManager.js
import { countOffset, defaultDensityFactor, defaultRemoveQuantity, deleteCount, double, empty, minCount, minIndex, minLimit, one, spatialHashGridCellSize, squareExp, } from "./Utils/Constants.js";
import { EventType } from "../Enums/Types/EventType.js";
import { LimitMode } from "../Enums/Modes/LimitMode.js";
import { Particle } from "./Particle.js";
import { SpatialHashGrid } from "./Utils/SpatialHashGrid.js";
import { getLogger } from "../Utils/LogUtils.js";
import { loadParticlesOptions } from "../Utils/OptionsUtils.js";
export class ParticlesManager {
    checkParticlePositionPlugins;
    grid;
    _array;
    _container;
    _groupLimits;
    _limit;
    _nextId;
    _particleBuckets;
    _particleResetPlugins;
    _particleUpdatePlugins;
    _pluginManager;
    _pool;
    _postParticleUpdatePlugins;
    _postUpdatePlugins;
    _resizeFactor;
    _updatePlugins;
    _zBuckets;
    constructor(pluginManager, container) {
        this._pluginManager = pluginManager;
        this._container = container;
        this._nextId = 0;
        this._array = [];
        this._pool = [];
        this._limit = 0;
        this._groupLimits = new Map();
        this._particleBuckets = new Map();
        this._zBuckets = this._createBuckets(this._container.zLayers);
        this.grid = new SpatialHashGrid(spatialHashGridCellSize);
        this.checkParticlePositionPlugins = [];
        this._particleResetPlugins = [];
        this._particleUpdatePlugins = [];
        this._postUpdatePlugins = [];
        this._postParticleUpdatePlugins = [];
        this._updatePlugins = [];
    }
    get count() {
        return this._array.length;
    }
    addParticle(position, overrideOptions, group, initializer) {
        const limitMode = this._container.actualOptions.particles.number.limit.mode, limit = group === undefined ? this._limit : (this._groupLimits.get(group) ?? this._limit), currentCount = this.count;
        if (limit > minLimit) {
            switch (limitMode) {
                case LimitMode.delete: {
                    const countToRemove = currentCount + countOffset - limit;
                    if (countToRemove > minCount) {
                        this.removeQuantity(countToRemove);
                    }
                    break;
                }
                case LimitMode.wait:
                    if (currentCount >= limit) {
                        return;
                    }
                    break;
                default:
                    break;
            }
        }
        try {
            const particle = this._pool.pop() ?? new Particle(this._pluginManager, this._container);
            particle.init(this._nextId, position, overrideOptions, group);
            let canAdd = true;
            if (initializer) {
                canAdd = initializer(particle);
            }
            if (!canAdd) {
                this._pool.push(particle);
                return;
            }
            this._array.push(particle);
            this._insertParticleIntoBucket(particle);
            this._nextId++;
            this._container.dispatchEvent(EventType.particleAdded, {
                particle,
            });
            return particle;
        }
        catch (e) {
            getLogger().warning(`error adding particle: ${e}`);
        }
        return undefined;
    }
    clear() {
        this._array = [];
        this._particleBuckets.clear();
        this._resetBuckets(this._container.zLayers);
    }
    destroy() {
        this._array = [];
        this._pool.length = 0;
        this._particleBuckets.clear();
        this._zBuckets = [];
        this.checkParticlePositionPlugins = [];
        this._particleResetPlugins = [];
        this._particleUpdatePlugins = [];
        this._postUpdatePlugins = [];
        this._postParticleUpdatePlugins = [];
        this._updatePlugins = [];
    }
    drawParticles(delta) {
        for (let i = this._zBuckets.length - one; i >= minIndex; i--) {
            const bucket = this._zBuckets[i];
            if (!bucket) {
                continue;
            }
            for (const particle of bucket) {
                particle.draw(delta);
            }
        }
    }
    filter(condition) {
        return this._array.filter(condition);
    }
    find(condition) {
        return this._array.find(condition);
    }
    get(index) {
        return this._array[index];
    }
    async init() {
        const container = this._container, options = container.actualOptions;
        this.checkParticlePositionPlugins = [];
        this._updatePlugins = [];
        this._particleUpdatePlugins = [];
        this._postUpdatePlugins = [];
        this._particleResetPlugins = [];
        this._postParticleUpdatePlugins = [];
        this._particleBuckets.clear();
        this._resetBuckets(container.zLayers);
        this.grid = new SpatialHashGrid(spatialHashGridCellSize * container.retina.pixelRatio);
        for (const plugin of container.plugins) {
            if (plugin.redrawInit) {
                await plugin.redrawInit();
            }
            if (plugin.checkParticlePosition) {
                this.checkParticlePositionPlugins.push(plugin);
            }
            if (plugin.update) {
                this._updatePlugins.push(plugin);
            }
            if (plugin.particleUpdate) {
                this._particleUpdatePlugins.push(plugin);
            }
            if (plugin.postUpdate) {
                this._postUpdatePlugins.push(plugin);
            }
            if (plugin.particleReset) {
                this._particleResetPlugins.push(plugin);
            }
            if (plugin.postParticleUpdate) {
                this._postParticleUpdatePlugins.push(plugin);
            }
        }
        await this._container.initDrawersAndUpdaters();
        for (const drawer of this._container.effectDrawers.values()) {
            await drawer.init?.(container);
        }
        for (const drawer of this._container.shapeDrawers.values()) {
            await drawer.init?.(container);
        }
        let handled = false;
        for (const plugin of container.plugins) {
            handled = plugin.particlesInitialization?.() ?? handled;
            if (handled) {
                break;
            }
        }
        if (!handled) {
            const particlesOptions = options.particles, groups = particlesOptions.groups;
            for (const group in groups) {
                const groupOptions = groups[group];
                if (!groupOptions) {
                    continue;
                }
                for (let i = this.count, j = 0; j < groupOptions.number.value && i < particlesOptions.number.value; i++, j++) {
                    this.addParticle(undefined, groupOptions, group);
                }
            }
            for (let i = this.count; i < particlesOptions.number.value; i++) {
                this.addParticle();
            }
        }
    }
    push(nb, position, overrideOptions, group) {
        for (let i = 0; i < nb; i++) {
            this.addParticle(position, overrideOptions, group);
        }
    }
    async redraw() {
        this.clear();
        await this.init();
        this._container.canvas.render.drawParticles({ value: 0, factor: 0 });
    }
    remove(particle, group, override) {
        this.removeAt(this._array.indexOf(particle), undefined, group, override);
    }
    removeAt(index, quantity = defaultRemoveQuantity, group, override) {
        if (index < minIndex || index > this.count) {
            return;
        }
        let deleted = 0;
        for (let i = index; deleted < quantity && i < this.count; i++) {
            if (this._removeParticle(i, group, override)) {
                i--;
                deleted++;
            }
        }
    }
    removeQuantity(quantity, group) {
        this.removeAt(minIndex, quantity, group);
    }
    setDensity() {
        const options = this._container.actualOptions, groups = options.particles.groups;
        let pluginsCount = 0;
        for (const plugin of this._container.plugins) {
            if (plugin.particlesDensityCount) {
                pluginsCount += plugin.particlesDensityCount();
            }
        }
        for (const group in groups) {
            const groupData = groups[group];
            if (!groupData) {
                continue;
            }
            const groupDataOptions = loadParticlesOptions(this._pluginManager, this._container, groupData);
            this._applyDensity(groupDataOptions, pluginsCount, group);
        }
        this._applyDensity(options.particles, pluginsCount);
    }
    setResizeFactor(factor) {
        this._resizeFactor = factor;
    }
    update(delta) {
        this.grid.clear();
        for (const plugin of this._updatePlugins) {
            plugin.update?.(delta);
        }
        const particlesToDelete = this._updateParticlesPhase1(delta);
        for (const plugin of this._postUpdatePlugins) {
            plugin.postUpdate?.(delta);
        }
        this._updateParticlesPhase2(delta, particlesToDelete);
        if (particlesToDelete.size) {
            for (const particle of particlesToDelete) {
                this.remove(particle);
            }
        }
        delete this._resizeFactor;
    }
    _addToPool = (...particles) => {
        this._pool.push(...particles);
    };
    _applyDensity = (options, pluginsCount, group, groupOptions) => {
        const numberOptions = options.number;
        if (!numberOptions.density.enable) {
            if (group === undefined) {
                this._limit = numberOptions.limit.value;
            }
            else if (groupOptions?.number.limit.value ?? numberOptions.limit.value) {
                this._groupLimits.set(group, groupOptions?.number.limit.value ?? numberOptions.limit.value);
            }
            return;
        }
        const densityFactor = this._initDensityFactor(numberOptions.density), optParticlesNumber = numberOptions.value, optParticlesLimit = numberOptions.limit.value > minLimit ? numberOptions.limit.value : optParticlesNumber, particlesNumber = Math.min(optParticlesNumber, optParticlesLimit) * densityFactor + pluginsCount, particlesCount = Math.min(this.count, this.filter(t => t.group === group).length);
        if (group === undefined) {
            this._limit = numberOptions.limit.value * densityFactor;
        }
        else {
            this._groupLimits.set(group, numberOptions.limit.value * densityFactor);
        }
        if (particlesCount < particlesNumber) {
            this.push(Math.abs(particlesNumber - particlesCount), undefined, options, group);
        }
        else if (particlesCount > particlesNumber) {
            this.removeQuantity(particlesCount - particlesNumber, group);
        }
    };
    _createBuckets = (zLayers) => {
        const bucketCount = Math.max(Math.floor(zLayers), one);
        return Array.from({ length: bucketCount }, () => []);
    };
    _getBucketIndex = (zIndex) => {
        const maxBucketIndex = this._zBuckets.length - one;
        if (maxBucketIndex <= minIndex) {
            return minIndex;
        }
        return Math.min(Math.max(Math.floor(zIndex), minIndex), maxBucketIndex);
    };
    _getParticleInsertIndex = (bucket, particleId) => {
        let start = minIndex, end = bucket.length;
        while (start < end) {
            const middle = Math.floor((start + end) / double), middleParticle = bucket[middle];
            if (!middleParticle) {
                end = middle;
                continue;
            }
            if (middleParticle.id < particleId) {
                start = middle + one;
            }
            else {
                end = middle;
            }
        }
        return start;
    };
    _initDensityFactor = densityOptions => {
        const container = this._container;
        if (!densityOptions.enable) {
            return defaultDensityFactor;
        }
        const canvasSize = container.canvas.size, pxRatio = container.retina.pixelRatio;
        if (!canvasSize.width || !canvasSize.height) {
            return defaultDensityFactor;
        }
        return ((canvasSize.width * canvasSize.height) / (densityOptions.height * densityOptions.width * pxRatio ** squareExp));
    };
    _insertParticleIntoBucket = (particle) => {
        const bucketIndex = this._getBucketIndex(particle.position.z), bucket = this._zBuckets[bucketIndex];
        if (!bucket) {
            return;
        }
        bucket.splice(this._getParticleInsertIndex(bucket, particle.id), empty, particle);
        this._particleBuckets.set(particle.id, bucketIndex);
    };
    _removeParticle = (index, group, override) => {
        const particle = this._array[index];
        if (!particle) {
            return false;
        }
        if (particle.group !== group) {
            return false;
        }
        this._array.splice(index, deleteCount);
        this._removeParticleFromBucket(particle);
        particle.destroy(override);
        this._container.dispatchEvent(EventType.particleRemoved, {
            particle,
        });
        this._addToPool(particle);
        return true;
    };
    _removeParticleFromBucket = (particle) => {
        const bucketIndex = this._particleBuckets.get(particle.id) ?? this._getBucketIndex(particle.position.z), bucket = this._zBuckets[bucketIndex];
        if (!bucket) {
            this._particleBuckets.delete(particle.id);
            return;
        }
        const particleIndex = this._getParticleInsertIndex(bucket, particle.id);
        if (bucket[particleIndex]?.id !== particle.id) {
            this._particleBuckets.delete(particle.id);
            return;
        }
        bucket.splice(particleIndex, deleteCount);
        this._particleBuckets.delete(particle.id);
    };
    _resetBuckets = (zLayers) => {
        const bucketCount = Math.max(Math.floor(zLayers), one);
        if (this._zBuckets.length !== bucketCount) {
            this._zBuckets = this._createBuckets(bucketCount);
            return;
        }
        for (const bucket of this._zBuckets) {
            bucket.length = minIndex;
        }
    };
    _updateParticleBucket = (particle) => {
        const newBucketIndex = this._getBucketIndex(particle.position.z), currentBucketIndex = this._particleBuckets.get(particle.id);
        if (currentBucketIndex === undefined) {
            this._insertParticleIntoBucket(particle);
            return;
        }
        if (currentBucketIndex === newBucketIndex) {
            return;
        }
        const currentBucket = this._zBuckets[currentBucketIndex];
        if (currentBucket) {
            const particleIndex = this._getParticleInsertIndex(currentBucket, particle.id);
            if (currentBucket[particleIndex]?.id === particle.id) {
                currentBucket.splice(particleIndex, deleteCount);
            }
        }
        const newBucket = this._zBuckets[newBucketIndex];
        if (!newBucket) {
            this._particleBuckets.set(particle.id, newBucketIndex);
            return;
        }
        newBucket.splice(this._getParticleInsertIndex(newBucket, particle.id), empty, particle);
        this._particleBuckets.set(particle.id, newBucketIndex);
    };
    _updateParticlesPhase1 = (delta) => {
        const particlesToDelete = new Set(), resizeFactor = this._resizeFactor;
        for (const particle of this._array) {
            if (resizeFactor && !particle.ignoresResizeRatio) {
                particle.position.x *= resizeFactor.width;
                particle.position.y *= resizeFactor.height;
                particle.initialPosition.x *= resizeFactor.width;
                particle.initialPosition.y *= resizeFactor.height;
            }
            particle.ignoresResizeRatio = false;
            for (const plugin of this._particleResetPlugins) {
                plugin.particleReset?.(particle);
            }
            for (const plugin of this._particleUpdatePlugins) {
                if (particle.destroyed) {
                    break;
                }
                plugin.particleUpdate?.(particle, delta);
            }
            if (particle.destroyed) {
                particlesToDelete.add(particle);
                continue;
            }
            this.grid.insert(particle);
        }
        return particlesToDelete;
    };
    _updateParticlesPhase2 = (delta, particlesToDelete) => {
        for (const particle of this._array) {
            if (particle.destroyed) {
                particlesToDelete.add(particle);
                continue;
            }
            for (const updater of this._container.particleUpdaters) {
                updater.update(particle, delta);
            }
            if (!particle.spawning) {
                for (const plugin of this._postParticleUpdatePlugins) {
                    plugin.postParticleUpdate?.(particle, delta);
                }
            }
            this._updateParticleBucket(particle);
        }
    };
}
