// @tsparticles/plugin-interactivity/tsparticles.plugin.interactivity.js
(function(g){g.__tsParticlesInternals=g.__tsParticlesInternals||{};g.__tsParticlesInternals.bundles=g.__tsParticlesInternals.bundles||{};g.__tsParticlesInternals.effects=g.__tsParticlesInternals.effects||{};g.__tsParticlesInternals.engine=g.__tsParticlesInternals.engine||{};g.__tsParticlesInternals.interactions=g.__tsParticlesInternals.interactions||{};g.__tsParticlesInternals.palettes=g.__tsParticlesInternals.palettes||{};g.__tsParticlesInternals.paths=g.__tsParticlesInternals.paths||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins=g.__tsParticlesInternals.plugins||{};g.__tsParticlesInternals.plugins.emittersShapes=g.__tsParticlesInternals.plugins.emittersShapes||{};g.__tsParticlesInternals.presets=g.__tsParticlesInternals.presets||{};g.__tsParticlesInternals.shapes=g.__tsParticlesInternals.shapes||{};g.__tsParticlesInternals.updaters=g.__tsParticlesInternals.updaters||{};g.__tsParticlesInternals.utils=g.__tsParticlesInternals.utils||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas=g.__tsParticlesInternals.canvas||{};g.__tsParticlesInternals.canvas.utils=g.__tsParticlesInternals.canvas.utils||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path=g.__tsParticlesInternals.path||{};g.__tsParticlesInternals.path.utils=g.__tsParticlesInternals.path.utils||{};var __tsProxyFactory=typeof Proxy!=="undefined"?function(obj){return new Proxy(obj,{get:function(target,key){if(!(key in target)){target[key]={};}return target[key];}});}:function(obj){return obj;};g.__tsParticlesInternals.bundles=__tsProxyFactory(g.__tsParticlesInternals.bundles);g.__tsParticlesInternals.effects=__tsProxyFactory(g.__tsParticlesInternals.effects);g.__tsParticlesInternals.interactions=__tsProxyFactory(g.__tsParticlesInternals.interactions);g.__tsParticlesInternals.palettes=__tsProxyFactory(g.__tsParticlesInternals.palettes);g.__tsParticlesInternals.paths=__tsProxyFactory(g.__tsParticlesInternals.paths);g.__tsParticlesInternals.plugins=__tsProxyFactory(g.__tsParticlesInternals.plugins);g.__tsParticlesInternals.plugins.emittersShapes=__tsProxyFactory(g.__tsParticlesInternals.plugins.emittersShapes);g.__tsParticlesInternals.presets=__tsProxyFactory(g.__tsParticlesInternals.presets);g.__tsParticlesInternals.shapes=__tsProxyFactory(g.__tsParticlesInternals.shapes);g.__tsParticlesInternals.updaters=__tsProxyFactory(g.__tsParticlesInternals.updaters);g.__tsParticlesInternals.utils=__tsProxyFactory(g.__tsParticlesInternals.utils);g.__tsParticlesInternals.canvas=__tsProxyFactory(g.__tsParticlesInternals.canvas);g.__tsParticlesInternals.path=__tsProxyFactory(g.__tsParticlesInternals.path);g.tsparticlesInternalExports=g.tsparticlesInternalExports||{};})(typeof globalThis!=="undefined"?globalThis:typeof window!=="undefined"?window:this);
/* Plugin v4.0.4 */
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@tsparticles/engine')) :
    typeof define === 'function' && define.amd ? define(['exports', '@tsparticles/engine'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.__tsParticlesInternals = global.__tsParticlesInternals || {}, global.__tsParticlesInternals.plugins = global.__tsParticlesInternals.plugins || {}, global.__tsParticlesInternals.plugins.interactivity = global.__tsParticlesInternals.plugins.interactivity || {}), global.__tsParticlesInternals.engine));
})(this, (function (exports, engine) { 'use strict';

    class ClickEvent {
        enable;
        mode;
        constructor() {
            this.enable = false;
            this.mode = [];
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.mode !== undefined) {
                this.mode = data.mode;
            }
        }
    }

    exports.DivType = void 0;
    (function (DivType) {
        DivType["circle"] = "circle";
        DivType["rectangle"] = "rectangle";
    })(exports.DivType || (exports.DivType = {}));

    class DivEvent {
        enable;
        mode;
        selectors;
        type;
        constructor() {
            this.selectors = [];
            this.enable = false;
            this.mode = [];
            this.type = exports.DivType.circle;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.selectors !== undefined) {
                this.selectors = data.selectors;
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.mode !== undefined) {
                this.mode = data.mode;
            }
            if (data.type !== undefined) {
                this.type = data.type;
            }
        }
    }

    class HoverEvent {
        enable;
        mode;
        constructor() {
            this.enable = false;
            this.mode = [];
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (data.enable !== undefined) {
                this.enable = data.enable;
            }
            if (data.mode !== undefined) {
                this.mode = data.mode;
            }
        }
    }

    class Events {
        onClick;
        onDiv;
        onHover;
        constructor() {
            this.onClick = new ClickEvent();
            this.onDiv = new DivEvent();
            this.onHover = new HoverEvent();
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            this.onClick.load(data.onClick);
            const onDiv = data.onDiv;
            if (onDiv !== undefined) {
                this.onDiv = engine.executeOnSingleOrMultiple(onDiv, t => {
                    const tmp = new DivEvent();
                    tmp.load(t);
                    return tmp;
                });
            }
            this.onHover.load(data.onHover);
        }
    }

    exports.InteractivityDetect = void 0;
    (function (InteractivityDetect) {
        InteractivityDetect["canvas"] = "canvas";
        InteractivityDetect["parent"] = "parent";
        InteractivityDetect["window"] = "window";
    })(exports.InteractivityDetect || (exports.InteractivityDetect = {}));

    class Modes {
        _container;
        _pluginManager;
        constructor(pluginManager, container) {
            this._pluginManager = pluginManager;
            this._container = container;
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            if (!this._container) {
                return;
            }
            const interactors = this._pluginManager.interactors?.get(this._container);
            if (!interactors) {
                return;
            }
            for (const interactor of interactors) {
                if (!interactor.loadModeOptions) {
                    continue;
                }
                interactor.loadModeOptions(this, data);
            }
        }
    }

    class Interactivity {
        detectsOn;
        events;
        modes;
        constructor(pluginManager, container) {
            this.detectsOn = exports.InteractivityDetect.window;
            this.events = new Events();
            this.modes = new Modes(pluginManager, container);
        }
        load(data) {
            if (engine.isNull(data)) {
                return;
            }
            const detectsOn = data.detectsOn;
            if (detectsOn !== undefined) {
                this.detectsOn = detectsOn;
            }
            this.events.load(data.events);
            this.modes.load(data.modes);
        }
    }

    class InteractivityPlugin {
        id = "interactivity";
        _pluginManager;
        constructor(pluginManager) {
            this._pluginManager = pluginManager;
        }
        async getPlugin(container) {
            const { InteractivityPluginInstance } = await Promise.resolve().then(function () { return InteractivityPluginInstance$1; });
            return new InteractivityPluginInstance(this._pluginManager, container);
        }
        loadOptions(container, options, source) {
            if (!this.needsPlugin()) {
                return;
            }
            let interactivityOptions = options.interactivity;
            if (!interactivityOptions?.load) {
                options.interactivity = interactivityOptions = new Interactivity(this._pluginManager, container);
            }
            interactivityOptions.load(source?.interactivity);
            const interactors = this._pluginManager.interactors?.get(container);
            if (!interactors) {
                return;
            }
            for (const interactor of interactors) {
                if (interactor.loadOptions) {
                    interactor.loadOptions(options, source);
                }
            }
        }
        loadParticlesOptions(container, options, source) {
            if (source?.interactivity) {
                options.interactivity = engine.deepExtend({}, source.interactivity);
            }
            const interactors = this._pluginManager.interactors?.get(container);
            if (!interactors) {
                return;
            }
            for (const interactor of interactors) {
                interactor.loadParticlesOptions?.(options, source);
            }
        }
        needsPlugin() {
            return true;
        }
    }

    exports.InteractorType = void 0;
    (function (InteractorType) {
        InteractorType["external"] = "external";
        InteractorType["particles"] = "particles";
    })(exports.InteractorType || (exports.InteractorType = {}));

    class ExternalInteractorBase {
        type = exports.InteractorType.external;
        container;
        constructor(container) {
            this.container = container;
        }
    }

    class ParticlesInteractorBase {
        type = exports.InteractorType.particles;
        container;
        constructor(container) {
            this.container = container;
        }
    }

    const clickEvent = "click", mouseDownEvent = "pointerdown", mouseUpEvent = "pointerup", mouseLeaveEvent = "pointerleave", mouseMoveEvent = "pointermove", touchStartEvent = "touchstart", touchEndEvent = "touchend", touchMoveEvent = "touchmove", touchCancelEvent = "touchcancel";

    function checkSelector(element, selectors) {
        const res = engine.executeOnSingleOrMultiple(selectors, selector => {
            return element.matches(selector);
        });
        return engine.isArray(res) ? res.some(t => t) : res;
    }
    function isDivModeEnabled(mode, divs) {
        return !!engine.findItemFromSingleOrMultiple(divs, t => t.enable && engine.isInArray(mode, t.mode));
    }
    function divModeExecute(mode, divs, callback) {
        engine.executeOnSingleOrMultiple(divs, div => {
            const divMode = div.mode, divEnabled = div.enable;
            if (divEnabled && engine.isInArray(mode, divMode)) {
                singleDivModeExecute(div, callback);
            }
        });
    }
    function singleDivModeExecute(div, callback) {
        const selectors = div.selectors;
        engine.executeOnSingleOrMultiple(selectors, selector => {
            callback(selector, div);
        });
    }
    function divMode(divs, element) {
        if (!element || !divs) {
            return;
        }
        return engine.findItemFromSingleOrMultiple(divs, div => {
            return checkSelector(element, div.selectors);
        });
    }

    async function loadInteractivityPlugin(engine$1) {
        engine$1.checkVersion("4.0.4");
        await engine$1.pluginManager.register(e => {
            const interactivityEngine = e, interactivityPluginManager = interactivityEngine.pluginManager;
            interactivityPluginManager.addPlugin(new InteractivityPlugin(interactivityPluginManager));
            interactivityPluginManager.initializers.interactors ??= new Map();
            interactivityPluginManager.interactors ??= new Map();
            interactivityPluginManager.addInteractor = (name, interactorInitializer) => {
                interactivityPluginManager.initializers.interactors ??= new Map();
                interactivityPluginManager.initializers.interactors.set(name, interactorInitializer);
            };
            interactivityPluginManager.getInteractors = async (container, force = false) => {
                interactivityPluginManager.interactors ??= new Map();
                interactivityPluginManager.initializers.interactors ??= new Map();
                return engine.getItemsFromInitializer(container, interactivityPluginManager.interactors, interactivityPluginManager.initializers.interactors, force);
            };
            interactivityPluginManager.setOnClickHandler = (callback) => {
                const { items } = interactivityEngine;
                if (!items.length) {
                    throw new Error("Click handlers can only be set after calling tsParticles.load()");
                }
                items.forEach(item => {
                    const interactivityContainer = item;
                    interactivityContainer.addClickHandler?.(callback);
                });
            };
        });
    }
    function ensureInteractivityPluginLoaded(e) {
        if (!e.pluginManager.addInteractor) {
            throw new Error("tsParticles Interactivity Plugin is not loaded");
        }
    }

    const globalObject = globalThis;
    globalObject.__tsParticlesInternals = globalObject.__tsParticlesInternals ?? {};
    globalObject.loadInteractivityPlugin = loadInteractivityPlugin;

    const touchDelay = 500;
    class InteractivityEventListeners {
        _canPush = true;
        _clickPositionPlugins;
        _container;
        _handlers;
        _interactionManager;
        _touches;
        constructor(container, interactionManager) {
            this._container = container;
            this._clickPositionPlugins = [];
            this._interactionManager = interactionManager;
            this._touches = new Map();
            this._handlers = {
                mouseDown: () => {
                    this._mouseDown();
                },
                mouseLeave: () => {
                    this._mouseTouchFinish();
                },
                mouseMove: (e) => {
                    this._mouseTouchMove(e);
                },
                mouseUp: (e) => {
                    this._mouseTouchClick(e);
                },
                touchStart: (e) => {
                    this._touchStart(e);
                },
                touchMove: (e) => {
                    this._mouseTouchMove(e);
                },
                touchEnd: (e) => {
                    this._touchEnd(e);
                },
                touchCancel: (e) => {
                    this._touchEnd(e);
                },
                touchEndClick: (e) => {
                    this._touchEndClick(e);
                },
                visibilityChange: () => {
                    this._handleVisibilityChange();
                },
            };
        }
        addListeners() {
            this._manageListeners(true);
        }
        init() {
            this._clickPositionPlugins.length = 0;
            for (const plugin of this._container.plugins.filter(p => !!p.clickPositionValid)) {
                this._clickPositionPlugins.push(plugin);
            }
        }
        removeListeners() {
            this._manageListeners(false);
        }
        _doMouseTouchClick = e => {
            const container = this._container, interactionManager = this._interactionManager, options = container.actualOptions;
            if (this._canPush) {
                const mouseInteractivity = interactionManager.interactivityData.mouse, mousePos = mouseInteractivity.position;
                if (!mousePos) {
                    return;
                }
                mouseInteractivity.clickPosition = { ...mousePos };
                mouseInteractivity.clickTime = performance.now();
                const onClick = options.interactivity?.events.onClick;
                if (!onClick?.mode) {
                    return;
                }
                engine.executeOnSingleOrMultiple(onClick.mode, mode => {
                    interactionManager.handleClickMode(mode);
                });
            }
            if (e.type === "touchend") {
                setTimeout(() => {
                    this._mouseTouchFinish();
                }, touchDelay);
            }
        };
        _handleVisibilityChange = () => {
            this._mouseTouchFinish();
        };
        _manageInteractivityListeners = add => {
            const handlers = this._handlers, container = this._container, interactionManager = this._interactionManager, options = container.actualOptions, interactivityEl = interactionManager.interactivityData.element;
            if (!interactivityEl) {
                return;
            }
            const html = interactivityEl, canvas = container.canvas;
            canvas.setPointerEvents(html === canvas.domElement ? "initial" : "none");
            if (add && !(options.interactivity?.events.onHover.enable || options.interactivity?.events.onClick.enable)) {
                return;
            }
            engine.manageListener(interactivityEl, mouseMoveEvent, handlers.mouseMove, add);
            engine.manageListener(interactivityEl, touchStartEvent, handlers.touchStart, add);
            engine.manageListener(interactivityEl, touchMoveEvent, handlers.touchMove, add);
            if (add) {
                if (options.interactivity?.events.onClick.enable) {
                    engine.manageListener(interactivityEl, touchEndEvent, handlers.touchEndClick, add);
                    engine.manageListener(interactivityEl, mouseUpEvent, handlers.mouseUp, add);
                    engine.manageListener(interactivityEl, mouseDownEvent, handlers.mouseDown, add);
                }
                else {
                    engine.manageListener(interactivityEl, touchEndEvent, handlers.touchEnd, add);
                }
            }
            else {
                engine.manageListener(interactivityEl, touchEndEvent, handlers.touchEndClick, add);
                engine.manageListener(interactivityEl, mouseUpEvent, handlers.mouseUp, add);
                engine.manageListener(interactivityEl, mouseDownEvent, handlers.mouseDown, add);
                engine.manageListener(interactivityEl, touchEndEvent, handlers.touchEnd, add);
            }
            engine.manageListener(interactivityEl, mouseLeaveEvent, handlers.mouseLeave, add);
            engine.manageListener(interactivityEl, touchCancelEvent, handlers.touchCancel, add);
        };
        _manageListeners = add => {
            const handlers = this._handlers, container = this._container, interactionManager = this._interactionManager, options = container.actualOptions, detectType = options.interactivity?.detectsOn, canvasEl = container.canvas.domElement;
            if (detectType === exports.InteractivityDetect.window) {
                interactionManager.interactivityData.element = engine.safeDocument();
            }
            else if (detectType === exports.InteractivityDetect.parent && canvasEl) {
                interactionManager.interactivityData.element = canvasEl.parentElement ?? canvasEl.parentNode;
            }
            else {
                interactionManager.interactivityData.element = canvasEl;
            }
            this._manageInteractivityListeners(add);
            engine.manageListener(document, engine.visibilityChangeEvent, handlers.visibilityChange, add, false);
        };
        _mouseDown = () => {
            const { interactivityData } = this._interactionManager, { mouse } = interactivityData;
            mouse.clicking = true;
            mouse.downPosition = mouse.position;
        };
        _mouseTouchClick = e => {
            const container = this._container, interactionManager = this._interactionManager, options = container.actualOptions, { mouse } = interactionManager.interactivityData;
            mouse.inside = true;
            let handled = false;
            const mousePosition = mouse.position;
            if (!mousePosition || !options.interactivity?.events.onClick.enable) {
                return;
            }
            for (const plugin of this._clickPositionPlugins) {
                handled = plugin.clickPositionValid?.(mousePosition) ?? false;
                if (handled) {
                    break;
                }
            }
            if (!handled) {
                this._doMouseTouchClick(e);
            }
            mouse.clicking = false;
        };
        _mouseTouchFinish = () => {
            const { interactivityData } = this._interactionManager, { mouse } = interactivityData;
            delete mouse.position;
            delete mouse.clickPosition;
            delete mouse.downPosition;
            interactivityData.status = mouseLeaveEvent;
            mouse.inside = false;
            mouse.clicking = false;
        };
        _mouseTouchMove = e => {
            const container = this._container, interactionManager = this._interactionManager, options = container.actualOptions, interactivity = interactionManager.interactivityData, canvasEl = container.canvas.domElement;
            if (!interactivity.element) {
                return;
            }
            interactivity.mouse.inside = true;
            let pos;
            if (e.type.startsWith("pointer")) {
                this._canPush = true;
                const mouseEvent = e;
                if (interactivity.element === engine.safeDocument()) {
                    if (canvasEl) {
                        const clientRect = canvasEl.getBoundingClientRect();
                        pos = {
                            x: mouseEvent.clientX - clientRect.left,
                            y: mouseEvent.clientY - clientRect.top,
                        };
                    }
                }
                else if (options.interactivity?.detectsOn === exports.InteractivityDetect.parent) {
                    const source = mouseEvent.target, target = mouseEvent.currentTarget;
                    if (canvasEl) {
                        const sourceRect = source.getBoundingClientRect(), targetRect = target.getBoundingClientRect(), canvasRect = canvasEl.getBoundingClientRect();
                        pos = {
                            x: mouseEvent.offsetX + engine.double * sourceRect.left - (targetRect.left + canvasRect.left),
                            y: mouseEvent.offsetY + engine.double * sourceRect.top - (targetRect.top + canvasRect.top),
                        };
                    }
                    else {
                        pos = {
                            x: mouseEvent.offsetX,
                            y: mouseEvent.offsetY,
                        };
                    }
                }
                else if (mouseEvent.target === canvasEl) {
                    pos = {
                        x: mouseEvent.offsetX,
                        y: mouseEvent.offsetY,
                    };
                }
            }
            else {
                this._canPush = e.type !== "touchmove";
                if (canvasEl) {
                    const touchEvent = e, lastTouch = touchEvent.touches[touchEvent.touches.length - engine.lengthOffset], canvasRect = canvasEl.getBoundingClientRect();
                    if (!lastTouch) {
                        return;
                    }
                    pos = {
                        x: lastTouch.clientX - canvasRect.left,
                        y: lastTouch.clientY - canvasRect.top,
                    };
                }
            }
            const pxRatio = container.retina.pixelRatio;
            if (pos) {
                pos.x *= pxRatio;
                pos.y *= pxRatio;
            }
            interactivity.mouse.position = pos;
            interactivity.status = mouseMoveEvent;
        };
        _touchEnd = e => {
            const evt = e, touches = Array.from(evt.changedTouches);
            for (const touch of touches) {
                this._touches.delete(touch.identifier);
            }
            this._mouseTouchFinish();
        };
        _touchEndClick = e => {
            const evt = e, touches = Array.from(evt.changedTouches);
            for (const touch of touches) {
                this._touches.delete(touch.identifier);
            }
            this._mouseTouchClick(e);
        };
        _touchStart = e => {
            const evt = e, touches = Array.from(evt.changedTouches);
            for (const touch of touches) {
                this._touches.set(touch.identifier, performance.now());
            }
            this._mouseTouchMove(e);
        };
    }

    const clickRadius = 1, touchEndLengthOffset = 1, minCoordinate = 0;
    class InteractionManager {
        container;
        interactivityData;
        _clickHandlers;
        _eventListeners;
        _externalInteractors;
        _interactors;
        _intersectionObserver;
        _particleInteractors;
        _pluginManager;
        constructor(pluginManager, container) {
            this.container = container;
            this._pluginManager = pluginManager;
            this._interactors = [];
            this._externalInteractors = [];
            this._particleInteractors = [];
            this._clickHandlers = new Map();
            this._eventListeners = new InteractivityEventListeners(container, this);
            this.interactivityData = {
                mouse: {
                    clicking: false,
                    inside: false,
                },
            };
            this._intersectionObserver = engine.safeIntersectionObserver(entries => {
                this._intersectionManager(entries);
            });
        }
        addClickHandler(callback) {
            const { container, interactivityData } = this;
            if (container.destroyed) {
                return;
            }
            const el = interactivityData.element;
            if (!el) {
                return;
            }
            const clickOrTouchHandler = (e, pos, radius) => {
                if (container.destroyed) {
                    return;
                }
                const pxRatio = container.retina.pixelRatio, posRetina = {
                    x: pos.x * pxRatio,
                    y: pos.y * pxRatio,
                }, particles = container.particles.grid.queryCircle(posRetina, radius * pxRatio);
                callback(e, particles);
            }, clickHandler = (e) => {
                if (container.destroyed) {
                    return;
                }
                const mouseEvent = e, pos = {
                    x: mouseEvent.offsetX,
                    y: mouseEvent.offsetY,
                };
                clickOrTouchHandler(e, pos, clickRadius);
            }, touchStartHandler = () => {
                if (container.destroyed) {
                    return;
                }
                touched = true;
                touchMoved = false;
            }, touchMoveHandler = () => {
                if (container.destroyed) {
                    return;
                }
                touchMoved = true;
            }, touchEndHandler = (e) => {
                if (container.destroyed) {
                    return;
                }
                if (touched && !touchMoved) {
                    const touchEvent = e, lastTouch = touchEvent.touches[touchEvent.touches.length - touchEndLengthOffset];
                    if (!lastTouch) {
                        return;
                    }
                    const element = container.canvas.domElement, canvasRect = element ? element.getBoundingClientRect() : undefined, pos = {
                        x: lastTouch.clientX - (canvasRect ? canvasRect.left : minCoordinate),
                        y: lastTouch.clientY - (canvasRect ? canvasRect.top : minCoordinate),
                    };
                    clickOrTouchHandler(e, pos, Math.max(lastTouch.radiusX, lastTouch.radiusY));
                }
                touched = false;
                touchMoved = false;
            }, touchCancelHandler = () => {
                if (container.destroyed) {
                    return;
                }
                touched = false;
                touchMoved = false;
            };
            let touched = false, touchMoved = false;
            this._clickHandlers.set(clickEvent, clickHandler);
            this._clickHandlers.set(touchStartEvent, touchStartHandler);
            this._clickHandlers.set(touchMoveEvent, touchMoveHandler);
            this._clickHandlers.set(touchEndEvent, touchEndHandler);
            this._clickHandlers.set(touchCancelEvent, touchCancelHandler);
            for (const [key, handler] of this._clickHandlers) {
                el.addEventListener(key, handler);
            }
        }
        addListeners() {
            this._eventListeners.addListeners();
        }
        clearClickHandlers() {
            const { container, interactivityData } = this;
            if (container.destroyed) {
                return;
            }
            for (const [key, handler] of this._clickHandlers) {
                interactivityData.element?.removeEventListener(key, handler);
            }
            this._clickHandlers.clear();
        }
        externalInteract(delta) {
            for (const interactor of this._externalInteractors) {
                const { interactivityData } = this;
                if (interactor.isEnabled(interactivityData)) {
                    interactor.interact(interactivityData, delta);
                }
            }
        }
        handleClickMode(mode) {
            if (this.container.destroyed) {
                return;
            }
            const { interactivityData } = this;
            for (const interactor of this._externalInteractors) {
                interactor.handleClickMode?.(mode, interactivityData);
            }
        }
        init() {
            this._eventListeners.init();
            for (const interactor of this._interactors) {
                switch (interactor.type) {
                    case exports.InteractorType.external:
                        this._externalInteractors.push(interactor);
                        break;
                    case exports.InteractorType.particles:
                        this._particleInteractors.push(interactor);
                        break;
                }
                interactor.init();
            }
        }
        async initInteractors() {
            const interactors = await this._pluginManager.getInteractors?.(this.container, true);
            if (!interactors) {
                return;
            }
            this._interactors = interactors;
            this._externalInteractors = [];
            this._particleInteractors = [];
        }
        particlesInteract(particle, delta) {
            const { interactivityData } = this;
            for (const interactor of this._externalInteractors) {
                interactor.clear(particle, delta);
            }
            for (const interactor of this._particleInteractors) {
                if (interactor.isEnabled(particle, interactivityData)) {
                    interactor.interact(particle, interactivityData, delta);
                }
            }
        }
        removeListeners() {
            this._eventListeners.removeListeners();
        }
        reset(particle) {
            const { interactivityData } = this;
            for (const interactor of this._externalInteractors) {
                if (interactor.isEnabled(interactivityData)) {
                    interactor.reset(interactivityData, particle);
                }
            }
            for (const interactor of this._particleInteractors) {
                if (interactor.isEnabled(particle, interactivityData)) {
                    interactor.reset(interactivityData, particle);
                }
            }
        }
        startObserving() {
            const { interactivityData } = this;
            if (interactivityData.element instanceof HTMLElement && this._intersectionObserver) {
                this._intersectionObserver.observe(interactivityData.element);
            }
        }
        stopObserving() {
            const { interactivityData } = this;
            if (interactivityData.element instanceof HTMLElement && this._intersectionObserver) {
                this._intersectionObserver.unobserve(interactivityData.element);
            }
        }
        updateMaxDistance() {
            let maxTotalDistance = 0;
            for (const interactor of this._interactors) {
                if (interactor.maxDistance > maxTotalDistance) {
                    maxTotalDistance = interactor.maxDistance;
                }
            }
            const container = this.container;
            container.particles.grid.setCellSize(maxTotalDistance * container.retina.pixelRatio);
        }
        _intersectionManager = entries => {
            const { container } = this;
            if (container.destroyed || !container.actualOptions.pauseOnOutsideViewport) {
                return;
            }
            for (const entry of entries) {
                if (entry.target !== this.interactivityData.element) {
                    continue;
                }
                if (entry.isIntersecting) {
                    container.play();
                }
                else {
                    container.pause();
                }
            }
        };
    }

    class InteractivityPluginInstance {
        interactionManager;
        _container;
        _pluginManager;
        constructor(pluginManager, container) {
            this._container = container;
            this._pluginManager = pluginManager;
            this.interactionManager = new InteractionManager(pluginManager, container);
            this._container.addClickHandler = (callback) => {
                this.interactionManager.addClickHandler(callback);
            };
        }
        addClickHandler(callback) {
            this.interactionManager.addClickHandler(callback);
        }
        clearClickHandlers() {
            this.interactionManager.clearClickHandlers();
        }
        destroy() {
            this.clearClickHandlers();
            this._pluginManager.interactors?.delete(this._container);
        }
        particleCreated(particle) {
            const interactivityParticle = particle, interactivity = new Interactivity(this._pluginManager, this._container);
            interactivity.load(this._container.actualOptions.interactivity);
            interactivity.load(interactivityParticle.options.interactivity);
            interactivityParticle.interactivity = interactivity;
        }
        particleReset(particle) {
            this.interactionManager.reset(particle);
        }
        postParticleUpdate(particle, delta) {
            this.interactionManager.particlesInteract(particle, delta);
        }
        postUpdate(delta) {
            this.interactionManager.externalInteract(delta);
            this.interactionManager.updateMaxDistance();
        }
        async preInit() {
            await this.interactionManager.initInteractors();
            this.interactionManager.init();
        }
        async redrawInit() {
            await this.interactionManager.initInteractors();
            this.interactionManager.init();
        }
        start() {
            this.interactionManager.addListeners();
            this.interactionManager.startObserving();
            return Promise.resolve();
        }
        stop() {
            this.interactionManager.removeListeners();
            this.interactionManager.stopObserving();
        }
    }

    var InteractivityPluginInstance$1 = /*#__PURE__*/Object.freeze({
        __proto__: null,
        InteractivityPluginInstance: InteractivityPluginInstance
    });

    exports.DivEvent = DivEvent;
    exports.ExternalInteractorBase = ExternalInteractorBase;
    exports.Modes = Modes;
    exports.ParticlesInteractorBase = ParticlesInteractorBase;
    exports.clickEvent = clickEvent;
    exports.divMode = divMode;
    exports.divModeExecute = divModeExecute;
    exports.ensureInteractivityPluginLoaded = ensureInteractivityPluginLoaded;
    exports.isDivModeEnabled = isDivModeEnabled;
    exports.loadInteractivityPlugin = loadInteractivityPlugin;
    exports.mouseDownEvent = mouseDownEvent;
    exports.mouseLeaveEvent = mouseLeaveEvent;
    exports.mouseMoveEvent = mouseMoveEvent;
    exports.mouseUpEvent = mouseUpEvent;
    exports.singleDivModeExecute = singleDivModeExecute;
    exports.touchCancelEvent = touchCancelEvent;
    exports.touchEndEvent = touchEndEvent;
    exports.touchMoveEvent = touchMoveEvent;
    exports.touchStartEvent = touchStartEvent;

}));
Object.assign(globalThis.window || globalThis, { loadInteractivityPlugin: (globalThis.__tsParticlesInternals.plugins.interactivity || {}).loadInteractivityPlugin });
delete (globalThis.window || globalThis).tsparticlesInternalExports;
