// @tsparticles/engine/cjs/Core/Utils/EventListeners.js
import { manageListener, safeDocument } from "../../Utils/Utils.js";
import { millisecondsToSeconds, resizeEvent, visibilityChangeEvent } from "./Constants.js";
export class EventListeners {
    container;
    _handlers;
    _resizeObserver;
    _resizeTimeout;
    constructor(container) {
        this.container = container;
        this._handlers = {
            visibilityChange: () => {
                this._handleVisibilityChange();
            },
            resize: () => {
                this._handleWindowResize();
            },
        };
    }
    addListeners() {
        this._manageListeners(true);
    }
    removeListeners() {
        this._manageListeners(false);
    }
    _handleVisibilityChange = () => {
        const container = this.container, options = container.actualOptions;
        if (!options.pauseOnBlur) {
            return;
        }
        if (safeDocument().hidden) {
            container.pageHidden = true;
            container.pause();
        }
        else {
            container.pageHidden = false;
            if (container.animationStatus) {
                container.play(true);
            }
            else {
                container.draw(true);
            }
        }
    };
    _handleWindowResize = () => {
        if (this._resizeTimeout) {
            clearTimeout(this._resizeTimeout);
            delete this._resizeTimeout;
        }
        const handleResize = async () => {
            const canvas = this.container.canvas;
            await canvas.windowResize();
        };
        this._resizeTimeout = setTimeout(() => void handleResize(), this.container.actualOptions.resize.delay * millisecondsToSeconds);
    };
    _manageListeners = add => {
        const handlers = this._handlers;
        this._manageResize(add);
        manageListener(document, visibilityChangeEvent, handlers.visibilityChange, add, false);
    };
    _manageResize = add => {
        const handlers = this._handlers, container = this.container, options = container.actualOptions;
        if (!options.resize.enable) {
            return;
        }
        if (typeof ResizeObserver === "undefined") {
            manageListener(globalThis, resizeEvent, handlers.resize, add);
            return;
        }
        const canvasEl = container.canvas.domElement;
        if (this._resizeObserver && !add) {
            if (canvasEl) {
                this._resizeObserver.unobserve(canvasEl);
            }
            this._resizeObserver.disconnect();
            delete this._resizeObserver;
        }
        else if (!this._resizeObserver && add && canvasEl) {
            this._resizeObserver = new ResizeObserver((entries) => {
                const entry = entries.find(e => e.target === canvasEl);
                if (!entry) {
                    return;
                }
                this._handleWindowResize();
            });
            this._resizeObserver.observe(canvasEl);
        }
    };
}
