// @tsparticles/engine/cjs/Core/Utils/SpatialHashGrid.js
import { Circle, Rectangle } from "./Ranges.js";
export class SpatialHashGrid {
    _cellSize;
    _cells = new Map();
    _circlePool = [];
    _circlePoolIdx;
    _pendingCellSize;
    _rectanglePool = [];
    _rectanglePoolIdx;
    constructor(cellSize) {
        this._cellSize = cellSize;
        this._circlePoolIdx = 0;
        this._rectanglePoolIdx = 0;
    }
    clear() {
        this._cells.clear();
        const pendingCellSize = this._pendingCellSize;
        if (pendingCellSize) {
            this._cellSize = pendingCellSize;
        }
        this._pendingCellSize = undefined;
    }
    insert(particle) {
        const { x, y } = particle.getPosition(), key = this._cellKeyFromCoords(x, y);
        if (!this._cells.has(key)) {
            this._cells.set(key, []);
        }
        this._cells.get(key)?.push(particle);
    }
    query(range, check, out = []) {
        const bounds = this._getRangeBounds(range);
        if (!bounds) {
            return out;
        }
        const minCellX = Math.floor(bounds.minX / this._cellSize), maxCellX = Math.floor(bounds.maxX / this._cellSize), minCellY = Math.floor(bounds.minY / this._cellSize), maxCellY = Math.floor(bounds.maxY / this._cellSize);
        for (let cx = minCellX; cx <= maxCellX; cx++) {
            for (let cy = minCellY; cy <= maxCellY; cy++) {
                const key = `${cx}_${cy}`, cellParticles = this._cells.get(key);
                if (!cellParticles) {
                    continue;
                }
                for (const p of cellParticles) {
                    if (check && !check(p)) {
                        continue;
                    }
                    if (range.contains(p.getPosition())) {
                        out.push(p);
                    }
                }
            }
        }
        return out;
    }
    queryCircle(position, radius, check, out = []) {
        const circle = this._acquireCircle(position.x, position.y, radius), result = this.query(circle, check, out);
        this._releaseShapes();
        return result;
    }
    queryRectangle(position, size, check, out = []) {
        const rect = this._acquireRectangle(position.x, position.y, size.width, size.height), result = this.query(rect, check, out);
        this._releaseShapes();
        return result;
    }
    setCellSize(cellSize) {
        this._pendingCellSize = cellSize;
    }
    _acquireCircle(x, y, r) {
        return (this._circlePool[this._circlePoolIdx++] ??= new Circle(x, y, r)).reset(x, y, r);
    }
    _acquireRectangle(x, y, w, h) {
        return (this._rectanglePool[this._rectanglePoolIdx++] ??= new Rectangle(x, y, w, h)).reset(x, y, w, h);
    }
    _cellKeyFromCoords(x, y) {
        const cellX = Math.floor(x / this._cellSize), cellY = Math.floor(y / this._cellSize);
        return `${cellX}_${cellY}`;
    }
    _getRangeBounds(range) {
        if (range instanceof Circle) {
            const r = range.radius, { x, y } = range.position;
            return {
                minX: x - r,
                maxX: x + r,
                minY: y - r,
                maxY: y + r,
            };
        }
        if (range instanceof Rectangle) {
            const { x, y } = range.position, { width, height } = range.size;
            return {
                minX: x,
                maxX: x + width,
                minY: y,
                maxY: y + height,
            };
        }
        return null;
    }
    _releaseShapes() {
        this._circlePoolIdx = 0;
        this._rectanglePoolIdx = 0;
    }
}
