// @tsparticles/plugin-emitters/esm/EmittersPluginInstance.js
import { isArray } from "@tsparticles/engine";
export class EmittersPluginInstance {
    container;
    _instancesManager;
    constructor(instancesManager, container) {
        this.container = container;
        this._instancesManager = instancesManager;
        this._instancesManager.initContainer(container);
    }
    async init() {
        const emittersOptions = this.container.actualOptions.emitters;
        if (isArray(emittersOptions)) {
            for (const emitterOptions of emittersOptions) {
                await this._instancesManager.addEmitter(this.container, emitterOptions);
            }
        }
        else {
            await this._instancesManager.addEmitter(this.container, emittersOptions);
        }
    }
    pause() {
        for (const emitter of this._instancesManager.getArray(this.container)) {
            emitter.pause();
        }
    }
    play() {
        for (const emitter of this._instancesManager.getArray(this.container)) {
            emitter.play();
        }
    }
    resize() {
        for (const emitter of this._instancesManager.getArray(this.container)) {
            emitter.resize();
        }
    }
    stop() {
        this._instancesManager.clear(this.container);
    }
    update(delta) {
        this._instancesManager.getArray(this.container).forEach(emitter => {
            emitter.update(delta);
        });
    }
}
